### PR TITLE
Implement LIMIT for pull queries

### PR DESCRIPTION
**Pull request by severb**
First pass at adding `LIMIT` (without `OFFSET`).

There are a couple of things I don't like about this approach:
* When increasing the limit new rows are not always added at the end of the previous listing. See the tests with LIMIT 5, 10, and 20 for an example. This is a result of the remote executor guesstimating the number of rows to request in parallel from each cluster, which interleaves new rows.
* We need to drain the pending rows which is wasteful, but otherwise [the old query will still be executing](/cashapp/pranadb/blob/main/pull/engine.go#L230).

Let's use this PR to decide if we want a better implementation in the short term. I have a few ideas where we could start.

There's also `PhysicalTopN` which we can simply model as a `PullLimit`->`PullSort`->`<whatever>` which comes with its own set of inefficiencies, but at least gets us going.

TODO:
- [x] Add unit tests

### CODE REVIEW

File: pull/exec_builder.go [hunk 1]
```diff
@@ -243,3 +246,41 @@ func (p *Engine) byItemsToDescAndSortExpression(byItems []*util.ByItems) ([]bool
 	}
 	return desc, sortByExprs
 }
+
```
Comment (severb) [inline pull/exec_builder.go:249]: I found these utilities handy to visualize the plans—we can exclude them if they're too distracting.

**Comment by purplefox** [comment]
> When increasing the limit new rows are not always added at the end of the previous listing. See the tests with LIMIT 5, 10, and 20 for an example. This is a result of the remote executor guesstimating the number of rows to request in parallel from each cluster, which interleaves new rows.

Ha, this is an interesting point. In general, any SQL query without an order by is not guaranteed to return the same rows every time it's executed. That's the case for pretty much every SQL database. So it's expected that increasing the limit in subsequent queries won't just return more rows than the previous executions of it. That's why pretty much all queries in other SQLTests include an order by - otherwise results would be non deterministic and tests would fail depending on the order of execution of the remote executors.

So we should definitely test with an order by clause there.

However! We also want to test without an order by, and that's going to be hard to do in a SQLTest because of the non determinism... Hmmm.

**Comment by purplefox** [comment]
> We need to drain the pending rows which is wasteful, but otherwise [the old query will still be executing](https://github.com/cashapp/pranadb/blob/main/pull/engine.go#L230).

I think you've discovered a bug in the current logic which determines when to remove the current remote query. Fixing the bug will mean we won't need to drain the pending rows.

One way of doing this would be to add a Close() method on the PullExecutor. When enough rows have been returned the Close can be called which propagates down the children and across the remote interface.

**Comment by purplefox** [comment]
> However! We also want to test without an order by, and that's going to be hard to do in a SQLTest because of the non determinism... Hmmm.

This can be done by having two columns in the source that we're testing against - one column has an ascending value like you currently do, and the other column has always the same value.

Then test limit without order by and only select the second column. The test will be deterministic since the values are the same, and the order doesn't matter anyway so the actual values don't matter.

Then test limit with order by and select the first column. This should be ordered correctly.

**Comment by purplefox** [comment]
Also, I'm not sure we should support offset at all. It's kind of useless without an order by anyway, as you'd be skipping random rows. The main reason for introducing limit is to stop the user executing queries that can return huge number of rows and take down the server - the idea being we will impose a strict limit on user's pull queries at some point.

**Comment by severb** [comment]
> One way of doing this would be to add a Close() method on the PullExecutor. When enough rows have been returned the Close can be called which propagates down the children and across the remote interface.

Adding a `Close()` method sounds good, I created #361 to track it. I think we can land this PR as is and follow-up with the improvement in another PR.

> Also, I'm not sure we should support offset at all. It's kind of useless without an order by anyway, as you'd be skipping random rows. The main reason for introducing limit is to stop the user executing queries that can return huge number of rows and take down the server - the idea being we will impose a strict limit on user's pull queries at some point.

Ok, I'll remove `OFFSET` for now; we can always add it back if we change our minds. Since for `PhysicalTopN` we need a `ORDER BY` + `LIMIT` and the largest sort size is currently 50k rows (`orderByMaxRows`), I'll enforce the same max on `LIMIT` too.

> Then test limit with order by and select the first column. This should be ordered correctly.

Neat idea, I'll give it a try.


I'll open a separate PR for `PhysicalTopN` and `ORDER BY` + `LIMIT` tests.

**Review by purplefox — APPROVED** [review]

**Comment by severb** [comment]
@purplefox can you please take another pass at f9865b5? I disabled `OFFSET`, added unit-test, and made the integration tests deterministic like you suggested. Next up it `PhysicalTopN` which should be straight-forward to do as a follow-up.

**Review by purplefox — APPROVED** [review]
LGTM

**Pull request merged**